source/lac/trilinos_tpetra_sparsity_pattern.cc

```diff
@@ -0,0 +1,1078 @@
+                          std::uint64_t(0)) <
+            static_cast<std::uint64_t>(std::numeric_limits<int>::max()),
+          ExcMessage("The TrilinosWrappers use Tpetra internally which "
+                     "uses 'signed int' to represent local indices. "
+                     "Therefore, only 2,147,483,647 nonzero matrix "
+                     "entries can be stored on a single process, "
+                     "but you are requesting more than that. "
+                     "If possible, use more MPI processes."));
+
+        if (row_map->getComm()->getSize() > 1)
+          graph = Utilities::Trilinos::internal::make_rcp<GraphType<Node>>(
+            row_map, n_entries_per_row());
+        else
+          graph = Utilities::Trilinos::internal::make_rcp<GraphType<Node>>(
+            row_map, col_map, n_entries_per_row());
+
+        AssertDimension(sp.n_rows(), graph->getGlobalNumEntries());
+
+        std::vector<TrilinosWrappers::types::int_type> row_indices;
+
+        // Include possibility to exchange data since DynamicSparsityPattern is
+        // able to do so
+        if (!exchange_data)
+          for (size_type<Node> row = first_row; row < last_row; ++row)
+            {
+              const TrilinosWrappers::types::int_type row_length =
+                sp.row_length(row);
+              if (row_length == 0)
+                continue;
+
+              row_indices.resize(row_length, -1);
+              {
+                typename SparsityPatternType::iterator p = sp.begin(row);
+                // avoid incrementing p over the end of the current row because
+                // it is slow for DynamicSparsityPattern in parallel
+                for (int col = 0; col < row_length;)
+                  {
+                    row_indices[col++] = p->column();
+                    if (col < row_length)
+                      ++p;
+                  }
+              }
+              graph->insertGlobalIndices(row, row_length, row_indices.data());
+            }
+        else
+          for (size_t row = 0; row < sp.n_rows(); ++row)
+            {
+              const TrilinosWrappers::types::int_type row_length =
+                sp.row_length(row);
+              if (row_length == 0)
+                continue;
+
+              row_indices.resize(row_length, -1);
+              {
+                typename SparsityPatternType::iterator p = sp.begin(row);
+                // avoid incrementing p over the end of the current row because
+                // it is slow for DynamicSparsityPattern in parallel
+                for (int col = 0; col < row_length;)
+                  {
+                    row_indices[col++] = p->column();
+                    if (col < row_length)
+                      ++p;
+                  }
+              }
+              const TrilinosWrappers::types::int_type trilinos_row = row;
+              graph->insertGlobalIndices(trilinos_row,
+                                         row_length,
```

Comment:
I don't see the difference between the `if` and `else` branches here, with the exception of the (likely unnecessary) cast of `row` to `trilinos_row` in line 461 (which is an artifact of the Epetra class). Are you sure this code is correct?

Reply:
See also #16410.

Reply:
You are right there is no difference.

The Epetra-based SparsityPattern uses an Epetra::FECrsGraph internally, but Tpetra has no equivalent. There is only the CrsGraph class in Tpetra.

If I understand the Epetra-based version correctly, the main benefit of the separation by cases is to gain a performance benefit in the case where `exchange_data == false`. (Somewhere in the Epetra-based class is a commend that it is fast to use  `Epetra_CrsGraph::InsertGlobalIndices` instead of  `Epetra_FECrsGraph::InsertGlobalIndices`.

I guess we can simply remove the separation by cases in the Tpetra case...

Reply:
Yes, please remove the separate cases if there is no difference.

Reply:
Done!